client/src/components/Forms/GenericForm.vue

```diff
@@ -0,0 +1,292 @@
+    title: {
+      type: String,
+      required: true,
+    },
+    model_key: {
+      type: String,
+      required: false,
+    },
+    params_key: {
+      type: String,
+      required: true,
+    },
+    modelChoices: {
+      type: Object,
+      required: true,
+    },
+    /**
+     * Holds the data for each model that could be chosen. The chosen model will
+     * switch what is in localHMModelFlat to what is held inside, and
+     * overwrite those values that correspond in localHMModelFlat.
+     */
+    modelChoicesData: {
+      type: Object,
+      required: true,
+    },
+    updateModelChoice: {
+      type: Function,
+      required: true,
+    },
+  },
+  components: {
+    DoubleField,
+    InputSlider,
+  },
+  data() {
+    return {
+      /**
+       * A local version of the relevant hmModelFlat.
+       */
+      localHMModelFlat: clonedeep(this.relevantHMModelFlat),
+      /**
+       * The cached options that the user has selected previously for the
+       * different models, but possibly haven't been saved to the server yet.
+       */
+      localModelChoicesData: clonedeep(this.modelChoicesData),
+    };
+  },
+  created() {
+    this.$watch(
+      function toWatch() {
+        return this.localHMModelFlat[this.model_key];
+      },
+      function updateParams(newModelName, oldModelName) {
+        if (oldModelName === newModelName) return;
+        // Use the method of updating appropriate to the form
+        const [newHMModelFlat, newModelChoicesData] = this.updateModelChoice(
+          oldModelName,
+          newModelName,
+          this.localHMModelFlat,
+          this.localModelChoicesData,
+        );
+        // Change everything but the model name to avoid infinite loop
+        delete newHMModelFlat[this.model_key];
+        this.localHMModelFlat = Object.assign(
+          this.localHMModelFlat,
+          newHMModelFlat,
+        );
+        this.localModelChoicesData = newModelChoicesData;
+      },
+    );
+  },
+  watch: {
+    /**
+     * If the state is changed locally, pass it up to the parent.
+     */
+    localHMModelFlat: {
+      deep: true,
+      handler() {
+        this.$emit('onChange', clonedeep(this.localHMModelFlat));
+      },
+    },
+    /**
+     * If the state is changed by the parent, change the local values.
+     */
+    relevantHMModelFlat: {
+      deep: true,
+      handler(newHMModelFlat) {
+        if (!isEqual(this.localHMModelFlat, newHMModelFlat)) {
+          this.localHMModelFlat = clonedeep(this.relevantHMModelFlat);
+        }
+      },
+    },
+  },
+  methods: {
+    getDoubleFieldProps(parameterKey) {
+      const parameterProps = PARAMETER_PROPS[parameterKey];
+      const doubleProps = {
+        param: parameterKey,
+        range: false,
+      };
+
+      // If properties do not exist, pass the default values
+      if (!parameterProps) return doubleProps;
+
+      if (parameterProps.plainName) {
+        doubleProps.param = parameterProps.plainName;
```

Comment:
is there any reason why this can't this be a simple spread operator? ```doubleProps = { ...parameterProps, ...doubleProps }```

Reply:
it would be nicer if doubleProps.param could be renamed to plainName or vice versa.

Reply:
Done! Great idea. That simplified things.